dbos/workflow.go

```diff
@@ -436,219 +537,238 @@ func runAsWorkflow[P any, R any](ctx context.Context, fn WorkflowFunc[P, R], inp
-				select {
-				case <-dbosWorkflowContext.Done():
-					// The context was cancelled or timed out: record timeout or cancellation
-					// CANCEL WORKFLOW HERE
-					return
-				}
-			}()
-		}
-	*/
+	// Commit the transaction
+	if err := tx.Commit(c.ctx); err != nil {
+		return nil, newWorkflowExecutionError(workflowID, fmt.Sprintf("failed to commit transaction: %v", err))
+	}
 
-	return handle, nil
+	return &workflowHandle[any]{workflowID: workflowID, outcomeChan: outcomeChan, dbosContext: c}, nil
 }
 
 /******************************/
 /******* STEP FUNCTIONS *******/
 /******************************/
 
-type StepFunc[P any, R any] func(ctx context.Context, input P) (R, error)
+type StepFunc func(ctx context.Context, input any) (any, error)
+type GenericStepFunc[P any, R any] func(ctx context.Context, input P) (R, error)
+
+const StepParamsKey DBOSContextKey = "stepParams"
 
 type StepParams struct {
 	MaxRetries    int
 	BackoffFactor float64
 	BaseInterval  time.Duration
 	MaxInterval   time.Duration
-}
-
-// stepOption is a functional option for configuring step parameters
-type stepOption func(*StepParams)
-
-// WithStepMaxRetries sets the maximum number of retries for a step
-func WithStepMaxRetries(maxRetries int) stepOption {
-	return func(p *StepParams) {
-		p.MaxRetries = maxRetries
+	StepName      string
+}
+
+// setStepParamDefaults returns a StepParams struct with all defaults properly set
+func setStepParamDefaults(params *StepParams, stepName string) *StepParams {
+	if params == nil {
+		return &StepParams{
+			MaxRetries:    0, // Default to no retries
+			BackoffFactor: 2.0,
+			BaseInterval:  100 * time.Millisecond, // Default base interval
+			MaxInterval:   5 * time.Second,        // Default max interval
+			StepName:      typeErasedStepNameToStepName[stepName],
+		}
 	}
-}
 
-// WithBackoffFactor sets the backoff factor for retries (multiplier for exponential backoff)
-func WithBackoffFactor(backoffFactor float64) stepOption {
-	return func(p *StepParams) {
-		p.BackoffFactor = backoffFactor
+	// Set defaults for zero values
+	if params.BackoffFactor == 0 {
+		params.BackoffFactor = 2.0 // Default backoff factor
 	}
-}
-
-// WithBaseInterval sets the base delay for the first retry
-func WithBaseInterval(baseInterval time.Duration) stepOption {
-	return func(p *StepParams) {
-		p.BaseInterval = baseInterval
+	if params.BaseInterval == 0 {
+		params.BaseInterval = 100 * time.Millisecond // Default base interval
+	}
+	if params.MaxInterval == 0 {
+		params.MaxInterval = 5 * time.Second // Default max interval
+	}
+	if params.StepName == "" {
+		// If the step name is not provided, use the function name
+		params.StepName = typeErasedStepNameToStepName[stepName]
 	}
+
+	return params
 }
 
-// WithMaxInterval sets the maximum delay for retries
-func WithMaxInterval(maxInterval time.Duration) stepOption {
-	return func(p *StepParams) {
-		p.MaxInterval = maxInterval
+var typeErasedStepNameToStepName = make(map[string]string)
```

Comment:
Again some pain:
- The user might or might not provide a custom step name
- The package level method knows the true name of the function
- The interface method sees only the typed erased wrapper name
- We want package-level and interface methods to have the same signature
- No functional options, no "step params" to smooth the step UX

--> We use an in-memory map to pass the function name to the interface method, which will use it for a default name if the user didn't provide any.